server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java

```diff
@@ -0,0 +1,294 @@
+ * would cover the case of tservers dying or coming online. To change the host pool check time period, set the following property: <br>
+ * <b>table.custom.balancer.host.regex.pool.check=5m</b><br>
+ * Regex matching can be based on either the host name (default) or host ip address. To set this balancer to match the regular expressions to the tablet server
+ * IP address, then set the following property:<br>
+ * <b>table.custom.balancer.host.regex.is.ip=true</b>
+ *
+ */
+public class HostRegexTableLoadBalancer extends TableLoadBalancer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HostRegexTableLoadBalancer.class);
+  public static final String HOST_BALANCER_PREFIX = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.";
+  public static final String HOST_BALANCER_OOB_CHECK = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.oob.period";
+  private static final String HOST_BALANCER_OOB_DEFAULT = "5m";
+  public static final String HOST_BALANCER_POOL_RECHECK_KEY = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.pool.check";
+  private static final String HOST_BALANCER_POOL_RECHECK_DEFAULT = "5m";
+  public static final String HOST_BALANCER_REGEX_USING_IPS = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.is.ip";
+  protected static final String DEFAULT_POOL = "HostTableLoadBalancer.ALL";
+
+  protected long oobCheckMillis = AccumuloConfiguration.getTimeInMillis(HOST_BALANCER_OOB_DEFAULT);
+  protected long poolRecheckMillis = AccumuloConfiguration.getTimeInMillis(HOST_BALANCER_POOL_RECHECK_DEFAULT);
+
+  private Map<String,String> tableIdToTableName = null;
+  private Map<String,Pattern> poolNameToRegexPattern = null;
+  private long lastOOBCheck = System.currentTimeMillis();
+  private long lastPoolRecheck = 0;
+  private boolean isIpBasedRegex = false;
+  private Map<String,SortedMap<TServerInstance,TabletServerStatus>> pools = new HashMap<String,SortedMap<TServerInstance,TabletServerStatus>>();
+
+  /**
+   * Group the set of current tservers by pool name. Tservers that don't match a regex are put into a default ppol.
+   *
+   * @param current
+   *          map of current tservers
+   * @return current servers grouped by pool name, if not a match it is put into a default pool.
+   */
+  protected synchronized Map<String,SortedMap<TServerInstance,TabletServerStatus>> splitCurrentByRegex(SortedMap<TServerInstance,TabletServerStatus> current) {
+    if ((System.currentTimeMillis() - lastPoolRecheck) > poolRecheckMillis) {
+      Map<String,SortedMap<TServerInstance,TabletServerStatus>> newPools = new HashMap<String,SortedMap<TServerInstance,TabletServerStatus>>();
+      for (Entry<TServerInstance,TabletServerStatus> e : current.entrySet()) {
+        List<String> poolNames = getPoolNamesForHost(e.getKey().host());
+        for (String pool : poolNames) {
+          if (!newPools.containsKey(pool)) {
```

Comment:
more efficient to do newPools.get(pool) and check result for null.  Avoids additional get down below when adding to the sorted map